### ISSUE TITLE
Should vector crypto instructions constraint vd_num/vs1_num/vs2_num align to lmul?

### ISSUE THREAD

**Issue by GuoShibo-cn** [riscv/riscv-isa-sim#1548]
![image](https://github.com/riscv-software-src/riscv-isa-sim/assets/117092168/70b9a723-80a8-48c8-8f62-f55f9417edc8)
According to zvkned_ext_macros.h, no where constraint vd/vs1 number align to lmul. And overlap just check vs_num == vd_num.
![image](https://github.com/riscv-software-src/riscv-isa-sim/assets/117092168/9f6686c8-4dbc-4ec0-83d9-2203d8a939d6)
however even with vlen =128, it would be possible to set lmul = 2, register groups could overlap with numbers not the same.

**Comment by scottj97** [comment]
Can you give a specific example of a case you think is broken today?

**Comment by GuoShibo-cn** [comment]
Sorry for delay.
![img_v3_026f_b9d5c7e7-d3df-4289-8ff2-c45445d54acg](https://github.com/riscv-software-src/riscv-isa-sim/assets/117092168/06f0dc19-af43-4cc2-8632-3d35fd380ede)
Our VLEN is 256 ,this is our simulation result, in the log we use lmul = 4, dst reg_num & src reg_num is not align to lmul, for vector  arithmatic instruction this would cause illegal instruction trap, howerver spike just continue the calcualtion.

**Comment by aswaterman** [comment]
cc @egouriou-rivos (and @chihminchao)

**Comment by timhsu404** [comment]
Yes, the vector crypto extensions should follow the LMUL alignment requirement.
Though the spec do not describe this behavior explicity, it does not make sense that register can not be the same but the content can still possible to overlap or out of the normal register range, the case v30 with LMUL=4 above.

I will create a PR for it.

**Comment by xinyuwang-starfive** [comment]
what status of the PR

**Comment by aswaterman** [comment]
cc @timhsu404

**Comment by nibrunieAtSi5** [comment]
@timhsu404 any progress on this issue ?

**Comment by chihminchao** [comment]
@timhsu404 is busy with another internal issue and may not have time on this issue this month.

**Comment by timhsu404** [comment]
> any progress on this issue ?

@nibrunieAtSi5 Not really. I have already done the very first version, but still tracking on some bugs.
The progress is pending for now, I'll keep working on this issue after the urgent job has been done.

**Comment by DavidYu360** [comment]
Vector crypto instruction VPR alignment rules:
Possible alignment value: `1, 2, 4, 8` (1 for any VPR, 2 for V0, V2, ...)
Prerequisite: `VLEN * LMUL >= EGW`
1. Single key in `.vs` inst: `max(EGW/VLEN, 1)`
2. Others: `max(LMUL, 1)`

**Comment by nibrunie** [comment]
I have developed a test case, https://gist.github.com/nibrunie/80a00047dce935011614530d86a829e6, which seems to be passing with flying colors on spike when I would have expected almost every instruction to trap.

**Comment by nibrunieAtSi5** [comment]
I think the issue is a bit worse than I initially thought: not checking alignment can also lead to segmentation fault as there is no guard to prevent spike from trying to access / write vector registers outside the valid `v0` ... `v31`.

For example, when selecting `vd=v31` and `LMUL > 1`.

The following program seems to seg fault on spike de5094a1a901d77ff44f89b38e00fefa15d4018e (Sept 20th):
```
#include <stdio.h>

// How to build:
//   riscv64-unknown-elf-gcc  -march=rv64gcv_zvkned vcrypto-lmul-alignment-test-case-seg-fault.c
// How to execute:
//   spike --isa=rv64gcv_zvbc_zicntr_zihpm_zvkned <pk-image> a.out

// This program aims at demonstrating that spike is flawed when
// it comes to executing vector crypto instructions with invalid register
// index alignments with respect to LMUL
int main(void) {
  asm volatile (
    // setting LMUL=4
    "vsetivli a1, 16, e32, m4, tu, mu\n"
    "vaesem.vv v20, v31\n" // <=== expected to trap but does not
    "vaesem.vv v31, v11\n" // <=== expected to trap but segfault
    :
    :
    : "a1"
  );
  printf("program should have trapped\n.");
  return 0;
}
```

**Comment by xinyuwang-starfive** [comment]
@nibrunie for example lmul=8 vaesdf.vs v8, v11 i think this  is overleap, but i think it will not trap in your PR?

**Comment by nibrunie** [comment]
> @nibrunie for example lmul=8 vaesdf.vs v8, v11 i think this is overleap, but i think it will not trap in your PR?

I think you are right @xinyuwang-starfive , this used to be covered by the alignment check (https://github.com/nibrunieAtSi5/riscv-isa-sim/blob/de5094a1a901d77ff44f89b38e00fefa15d4018e/riscv/zvkned_ext_macros.h#L21) but since we are implementing the proper relaxed index constraint for the scalar operand the alignment check is no longer sufficient.

I will try to have a proper / closer look over the week-end and fix my PRs. 
Thank you for spotting this @xinyuwang-starfive

**Comment by xinyuwang-starfive** [comment]
@nibrunie 
 in vghsh.vv it only check  vs2 and vd , i think the vs1 is also need to check align ?
![image](https://github.com/user-attachments/assets/7616502d-a0da-4c95-92dd-db64ba46ad5b)

**Comment by nibrunie** [comment]
> @nibrunie in vghsh.vv it only check vs2 and vd , i think the vs1 is also need to check align ? ![image](https://private-user-images.githubusercontent.com/37825795/376487449-7616502d-a0da-4c95-92dd-db64ba46ad5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkzNjIwODIsIm5iZiI6MTcyOTM2MTc4MiwicGF0aCI6Ii8zNzgyNTc5NS8zNzY0ODc0NDktNzYxNjUwMmQtYTBkYS00Yzk1LTkyZGQtZGI2NGJhNDZhZDViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE5VDE4MTYyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlNzNjNTc4MzI3ZjQ4NzYxZGE2OTRkNDVjZTRiNDIxZDhmYjhjZGQ4MzAwN2NmN2Y4NDMwZDJiZDcxMzBjYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Q08tX36U7meiDetVY6NX8DQkSnq8w1VItG5DGkkCzJ0)

@xinyuwang-starfive , this was fixed on Sept 25th (https://github.com/riscv-software-src/riscv-isa-sim/commit/b1ce0a35365d9a8508f02f1ba2dc5b339f892d84) on the PR opened to address this issue https://github.com/riscv-software-src/riscv-isa-sim/pull/1815. Is your branch up-to-date ?

**Comment by MrSljuka** [comment]
Hi All, is this issue still open?

**Comment by nibrunieAtSi5** [comment]
This issue should be closed, it is supposed to have been fixed by https://github.com/riscv-software-src/riscv-isa-sim/commit/b30b11dfefe27c7c4f709df991bd91d5b6c9464c
